File: results/dnd-rest-benchmark/lifecycle-runs/20260730T082348Z_claude_claude-sonnet-5_go-stdlib/play_invitations.go
package main

import (
	"encoding/json"
	"net/http"
	"strings"
)

// playInvitation is a campaign dm's invitation for a registered player
// identity to join the campaign with a specific character id.
type playInvitation struct {
	InvitationID string
	Username     string
	CharacterID  string
	Status       string
}

func playInvitationResponse(inv *playInvitation) map[string]interface{} {
	return map[string]interface{}{
		"invitation_id": inv.InvitationID,
		"username":      inv.Username,
		"character_id":  inv.CharacterID,
		"status":        inv.Status,
	}
}

// handlePlayCampaignInvitationsSub routes the "invitations" and
// "invitations/{id}/accept" sub-paths of a play campaign. It returns false
// if rest does not name a recognized invitations path, so the caller can
// fall through to its own routing.
func handlePlayCampaignInvitationsSub(w http.ResponseWriter, r *http.Request, campaignID, rest string) bool {
	if rest == "invitations" {
		switch r.Method {
		case http.MethodPost:
			handleCreatePlayInvitation(w, r, campaignID)
		case http.MethodGet:
			handleListPlayInvitations(w, r, campaignID)
		default:
			writeError(w, http.StatusMethodNotAllowed, "method not allowed")
		}
		return true
	}
	if invRest, ok := strings.CutPrefix(rest, "invitations/"); ok && invRest != "" {
		if invID, ok := strings.CutSuffix(invRest, "/accept"); ok && invID != "" {
			if r.Method != http.MethodPost {
				writeError(w, http.StatusMethodNotAllowed, "method not allowed")
				return true
			}
			handleAcceptPlayInvitation(w, r, campaignID, invID)
			return true
		}
	}
	return false
}

type playInvitationRequest struct {
	InvitationID string `json:"invitation_id"`
	Username     string `json:"username"`
	CharacterID  string `json:"character_id"`
}

// handleCreatePlayInvitation lets the campaign dm invite a registered player
// identity to join with a specific character id.
func handleCreatePlayInvitation(w http.ResponseWriter, r *http.Request, campaignID string) {
	username, ok := authenticatePlay(r)
	if !ok {
		writeError(w, http.StatusUnauthorized, "authentication required")
		return
	}

	var req playInvitationRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}

	playMu.Lock()
	c, ok := requirePlayCampaign(w, campaignID)
	if !ok {
		return
	}
	if c.Owner != username {
		playMu.Unlock()
		writeError(w, http.StatusForbidden, "only the campaign dm may create invitations")
		return
	}
	if req.InvitationID == "" || req.Username == "" || req.CharacterID == "" {
		playMu.Unlock()
		writeError(w, http.StatusBadRequest, "invitation_id, username, and character_id are required")
		return
	}
	target := lookupPlayAccount(req.Username)
	if target == nil || target.Role != "player" {
		playMu.Unlock()
		writeError(w, http.StatusBadRequest, "username must name a registered player")
		return
	}
	if c.Invitations == nil {
		c.Invitations = make(map[string]*playInvitation)
	}
	if _, exists := c.Invitations[req.InvitationID]; exists {
		playMu.Unlock()
		writeError(w, http.StatusConflict, "invitation_id already exists")
		return
	}
	for _, inv := range c.Invitations {
		if inv.Username == req.Username && inv.Status == "pending" {
			playMu.Unlock()
			writeError(w, http.StatusConflict, "this user already has a pending invitation")
			return
		}
	}

	rec := &playInvitation{
		InvitationID: req.InvitationID,
		Username:     req.Username,
		CharacterID:  req.CharacterID,
		Status:       "pending",
	}
	c.Invitations[req.InvitationID] = rec
	c.InvitationOrder = append(c.InvitationOrder, req.InvitationID)
	resp := playInvitationResponse(rec)
	playMu.Unlock()
	persistState()

	writeJSON(w, http.StatusCreated, resp)
}

// handleListPlayInvitations returns campaign invitations in creation order.
// The campaign dm sees all invitations; a target user sees only their own
// invitations (even before becoming a member), and everyone else sees none.
func handleListPlayInvitations(w http.ResponseWriter, r *http.Request, campaignID string) {
	username, ok := authenticatePlay(r)
	if !ok {
		writeError(w, http.StatusUnauthorized, "authentication required")
		return
	}

	playMu.Lock()
	c, ok := requirePlayCampaign(w, campaignID)
	if !ok {
		return
	}

	isDM := c.Owner == username
	invitations := make([]map[string]interface{}, 0, len(c.InvitationOrder))
	for _, invID := range c.InvitationOrder {
		rec := c.Invitations[invID]
		if isDM || rec.Username == username {
			invitations = append(invitations, playInvitationResponse(rec))
		}
	}
	playMu.Unlock()

	writeJSON(w, http.StatusOK, map[string]interface{}{"invitations": invitations})
}

// handleAcceptPlayInvitation lets only the invited target user accept a
// pending invitation, adding them as a campaign member with the invitation's
// character id.
func handleAcceptPlayInvitation(w http.ResponseWriter, r *http.Request, campaignID, invitationID string) {
	username, ok := authenticatePlay(r)
	if !ok {
		writeError(w, http.StatusUnauthorized, "authentication required")
		return
	}

	playMu.Lock()
	c, ok := requirePlayCampaign(w, campaignID)
	if !ok {
		return
	}
	rec := c.Invitations[invitationID]
	if rec == nil {
		playMu.Unlock()
		writeError(w, http.StatusNotFound, "invitation not found")
		return
	}
	if rec.Username != username {
		playMu.Unlock()
		writeError(w, http.StatusForbidden, "only the invited user may accept this invitation")
		return
	}
	if rec.Status == "accepted" {
		playMu.Unlock()
		writeError(w, http.StatusConflict, "invitation already accepted")
		return
	}

	m := &playMember{
		Username:    username,
		CharacterID: rec.CharacterID,
		HPCurrent:   20,
		HPMax:       20,
		Owner:       username,
		Gold:        10,
	}
	c.Members = append(c.Members, m)
	rec.Status = "accepted"
	resp := playInvitationResponse(rec)
	playMu.Unlock()
	persistState()

	writeJSON(w, http.StatusOK, resp)
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260730T075550Z_pi_kimi-k2p7-code_ruby-sinatra/lib/routes/play_invitations.rb
# frozen_string_literal: true

# Route definitions for campaign invitation endpoints.

# --- Create invitation ---

post '/v1/play/campaigns/:id/invitations' do
  content_type :json
  campaign_id = params[:id]
  validate_campaign_id!(campaign_id)

  username = require_dm_actor!
  campaign = load_play_campaign!(campaign_id)
  require_play_campaign_owner!(campaign, username)

  body = parse_json_body
  validate_invitation_body!(body)

  target_user = Storage.load_user(body['username'])
  json_error(400, 'invalid target user') unless target_user && target_user[:role] == 'player'

  json_error(409, 'invitation already exists') if Storage.invitation_exists?(campaign_id, body['invitation_id'])
  json_error(409, 'pending invitation already exists for user') if Storage.pending_invitation_exists?(campaign_id, body['username'])

  Storage.create_invitation(campaign_id, body['invitation_id'], body['username'], body['character_id'])

  status 201
  JSON.dump(
    invitation_id: body['invitation_id'],
    username: body['username'],
    character_id: body['character_id'],
    status: 'pending'
  )
end

# --- Accept invitation ---

post '/v1/play/campaigns/:id/invitations/:invitation_id/accept' do
  content_type :json
  campaign_id = params[:id]
  validate_campaign_id!(campaign_id)

  actor = authenticate_actor!
  username = actor[:username]
  campaign = load_play_campaign!(campaign_id)

  invitation_id = params[:invitation_id]
  validate_invitation_id!(invitation_id)

  invitation = Storage.load_invitation(campaign_id, invitation_id)
  json_error(404, 'invitation not found') unless invitation

  json_error(403, 'forbidden') unless invitation[:username] == username

  json_error(409, 'invitation already accepted') if invitation[:status] == 'accepted'

  unless Storage.play_campaign_member_exists?(campaign_id, username)
    json_error(409, 'character already exists') if Storage.play_campaign_character_exists?(campaign_id, invitation[:character_id])

    Storage.create_play_campaign_member(campaign_id, username, invitation[:character_id], invitation[:character_id], 'fighter')
  end

  Storage.accept_invitation(campaign_id, invitation_id)

  status 200
  JSON.dump(
    invitation_id: invitation[:invitation_id],
    username: invitation[:username],
    character_id: invitation[:character_id],
    status: 'accepted'
  )
end

# --- List invitations ---

get '/v1/play/campaigns/:id/invitations' do
  content_type :json
  campaign_id = params[:id]
  validate_campaign_id!(campaign_id)

  actor = authenticate_actor!
  username = actor[:username]
  campaign = load_play_campaign!(campaign_id)

  invitations = if campaign[:owner] == username
                  Storage.load_invitations(campaign_id)
                else
                  own = Storage.load_invitations_for_user(campaign_id, username)
                  if !own.empty?
                    own
                  elsif Storage.play_campaign_member_exists?(campaign_id, username)
                    []
                  else
                    json_error(403, 'forbidden')
                  end
                end

  JSON.dump(
    invitations: invitations.map do |inv|
      {
        invitation_id: inv[:invitation_id],
        username: inv[:username],
        character_id: inv[:character_id],
        status: inv[:status]
      }
    end
  )
end
